src/accountsdb/db.zig

```diff
@@ -1354,179 +1327,146 @@ pub const AccountsDB = struct {
-                    }
-                } else {
-                    // hashes aren't always stored correctly in snapshots
-                    if (account_hash.eql(Hash.ZEROES)) {
-                        const account, var account_lg =
-                            try self.getAccountFromRefWithReadLock(max_slot_ref);
-                        defer {
-                            account_lg.unlock();
-                            switch (account) {
-                                .file => |in_file_account| in_file_account.deinit(self.allocator),
-                                .unrooted_map => {},
-                            }
-                        }
+                const ref_head = key.value_ptr;
 
-                        account_hash = switch (account) {
-                            .file => |in_file_account| blk: {
-                                const info = in_file_account.account_info;
-                                var iterator = in_file_account.data.iterator();
-                                var hash = Hash.ZEROES;
-                                sig.core.account.hashAccount(
-                                    info.lamports,
-                                    &iterator,
-                                    &info.owner,
-                                    info.executable,
-                                    info.rent_epoch,
-                                    &in_file_account.store_info.pubkey,
-                                    &hash.data,
-                                );
-                                break :blk hash;
-                            },
-                            .unrooted_map => |unrooted_account| unrooted_account.hash(key),
-                        };
-                    }
-                }
+                // "mix in" account into hash
+                const latest_slot = slot: {
+                    // get the most recent state of the account
+                    const max_slot_ref = slotListMaxWithinBounds(
+                        ref_head.ref_ptr,
+                        maybe_min_slot,
+                        maybe_max_slot,
+                    ) orelse continue;
 
-                hashes.appendAssumeCapacity(account_hash);
-                local_total_lamports += lamports;
+                    const account = try self.getAccountFromRef(allocator, max_slot_ref);
+                    defer account.deinit(allocator);
+
+                    lt_hash.mixIn(account.ltHash(max_slot_ref.pubkey));
+
+                    total_lamports += account.lamports;
+
+                    if (max_slot_ref.slot == 0) continue;
+                    break :slot max_slot_ref.slot;
+                };
+
+                // "mix out" old version of account from hash
+                {
+                    // I expect this branch to not be taken for full snapshots, however it would
+                    // still be entirely legal.
```

Comment:
This is not correct if you're calculating an lthash over the entire database from scratch. The lthash of the database can be calculated by mixing every account's hash into identity. Mixing out old versions from that will result in an incorect lthash.

Reply:
I believe the lthash should be a hash of all *accounts* at the slot we're concerned with, not all entries for each account.

Any entry for an account which has a later entry can be considered overwritten - if a snapshot contains old entries I think this is legal but basically junk data